docs/storage_transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,185 @@
+Attribution 3.0 Unported
+License <https://creativecommons.org/licenses/by/3.0/>`_.
+
+----
+
+
+Abstract
+========
+
+This specification defines an implementation of the Zarr
+storage transformer protocol for sharding.
+
+
+Motivation
+==========
+
+Sharding decouples the concept of chunks from storage keys, which become shards.
+This is helpful when the requirements for those don't align:
+
+- Compressible units of chunks often need to be read and written in smaller
+  chunks, whereas
+- storage often is optimized for larger data per entry and fewer entries, e.g.
+  as restricted by the file block size and maximum inode number for typical
+  file systems.
+
+This does not necessarily fit the access patterns of the data, so chunks might
+need to be smaller than one storage key. In those cases sharding decouples those
+entities. One shard corresponds to one storage key, but can contain multiple chunks:
+
+.. image:: sharding.png
+
+
+Document conventions
+====================
+
+Conformance requirements are expressed with a combination of
+descriptive assertions and [RFC2119]_ terminology. The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in the normative
+parts of this document are to be interpreted as described in
+[RFC2119]_. However, for readability, these words do not appear in all
+uppercase letters in this specification.
+
+All of the text of this specification is normative except sections
+explicitly marked as non-normative, examples, and notes. Examples in
+this specification are introduced with the words "for example".
+
+
+Configuration
+=============
+
+Sharding can be configured per array in the :ref:`array-metadata`:
+
+.. code-block::
+
+    {
+      storage_transformers: [
+        {
+          "storage_transformer": "https://purl.org/zarr/spec/storage_transformers/sharding/1.0",
+          "configuration": {
+            "format": "indexed",
+            "chunks_per_shard": [
+                2,
+                2
+            ]
+          }
+      ]
+    }
+
+``format``
+
+    Specifies a `Binary shard format`_. In this version, the only binary format is the
+    ``indexed`` format.
+
+``chunks_per_shard``
+
+    An array of integers providing the number of chunks that are combined in a shard
+    for each dimension of the Zarr array, where each chunk may only start at a position
+    that is divisble by ``chunks_per_shard`` per dimension, e.g. starting at the zero-origin.
+    The length of the array must match the length of the array metadata ``shape`` entry.
+    For example, a value ``[32, 2]`` indicates that 64 chunks are combined in one shard,
+    32 along the first dimension, and for each of those 2 along the second dimension.
+    Valid starting positions for a shard in the chunk-grid are therefore ``[0, 0]``,
+    ``[32, 2]``, ``[32, 4]``, ``[64, 2]`` or ``[96, 18]``.
+
+
+Key & value transformation
```

Comment:
This is basically the storage transformer implementation for sharding, right?
Maybe reflect that in this heading.

Reply:
Key & value transformation with sharding storage transformer?

Reply:
Instead I added `Storage transformer implementation` as another heading and put the other two as subsections:

```
Storage transformer implementation
==================================

Key & value transformation
--------------------------

…


Binary shard format
-------------------

…
```